js/private/js_library.bzl

```diff
@@ -0,0 +1,106 @@
+"""Expose some files with DeclarationInfo, like filegroup but can be a dep of ts_project
+"""
+
+load("@aspect_bazel_lib//lib:copy_to_bin.bzl", "copy_files_to_bin_actions")
+load("@rules_nodejs//nodejs:providers.bzl", "DeclarationInfo", "declaration_info")
+
+_DOC = """Copies all sources to the output tree and expose some files with DeclarationInfo.
+
+Can be used as a dep for rules that expect a DeclarationInfo such as ts_project."""
+
+_ATTRS = {
+    "srcs": attr.label_list(allow_files = True),
+    "deps": attr.label_list(allow_files = True),
+    "_windows_constraint": attr.label(default = "@platforms//os:windows"),
+}
+
+def _js_library_impl(ctx):
+    is_windows = ctx.target_platform_has_constraint(ctx.attr._windows_constraint[platform_common.ConstraintValueInfo])
+
+    all_files = []
+    typings = []
+
+    for file in ctx.files.srcs:
+        if ctx.label.package != file.owner.package:
+            msg = """
+
+Expected to find file {file_basename} in {this_package}, but instead it is in {file_package}.
+
+All srcs in a js_library must be in the same package as the js_library target.
+
+Either move {file_basename} to {this_package}, or create a js_library
+target in {file_basename}'s package and add that target to the deps of {this_target}:
+
+    buildozer 'new_load @aspect_rules_js//js:defs.bzl js_library' {file_package}:__pkg__
+    buildozer 'new js_library {new_target_name}' {file_package}:__pkg__
+    buildozer 'add srcs {file_basename}' {file_package}:{new_target_name}
+    buildozer 'add visibility {this_package}:__pkg__' {file_package}:{new_target_name}
+    buildozer 'remove srcs {file_package}:{file_basename}' {this_target}
+    buildozer 'add deps {file_package}:{new_target_name}' {this_target}
+
+""".format(
+                file_basename = file.basename,
+                file_package = "%s//%s" % (file.owner.workspace_name, file.owner.package),
+                new_target_name = file.basename.replace(".", "_"),
+                this_package = "%s//%s" % (ctx.label.workspace_name, ctx.label.package),
+                this_target = ctx.label,
+            )
+            fail(msg)
+
+    output_srcs = copy_files_to_bin_actions(ctx, ctx.files.srcs, is_windows = is_windows)
+
+    for src in output_srcs:
+        all_files.append(src)
```

Comment:
nit: no need to create `all_files` as its always identical to `ctx.files.srcs`

Reply:
👍